File: mobile/src/test/java/org/openhab/habdroid/util/SyncHttpClientTest.java
package org.openhab.habdroid.util;

import okhttp3.OkHttpClient;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.powermock.core.classloader.annotations.PowerMockIgnore;
import org.powermock.modules.junit4.PowerMockRunner;

import java.net.UnknownHostException;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

@RunWith(PowerMockRunner.class)
@PowerMockIgnore("javax.net.ssl.*")
public class SyncHttpClientTest {
    OkHttpClient mClient;

    @Before
    public void setupClient() {
        mClient = new OkHttpClient.Builder().build();
    }

    /**
     * Unit test against Issue #315 "Crash when connection could not be established".
     */
    @Test
    public void testMethodErrorResponse() {
        SyncHttpClient httpClient = new SyncHttpClient(mClient, "https://demo.test",
                null, null);

        String host = "just.a.local.url.local";
        SyncHttpClient.HttpStatusResult resp = httpClient.get("https://" + host).asStatus();

        assertEquals(500, resp.statusCode);
        assertTrue(resp.error instanceof UnknownHostException);
    }
}

File: mobile/src/test/java/org/openhab/habdroid/util/SyncHttpClientTest.kt
/*
 * Copyright (c) 2010-2020 Contributors to the openHAB project
 *
 * See the NOTICE file(s) distributed with this work for additional
 * information.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0
 *
 * SPDX-License-Identifier: EPL-2.0
 */

package org.openhab.habdroid.util

import java.net.UnknownHostException
import kotlinx.coroutines.runBlocking
import okhttp3.OkHttpClient
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test

class SyncHttpClientTest {
    private lateinit var client: OkHttpClient

    @Before
    fun setupClient() {
        client = OkHttpClient.Builder().build()
    }

    /**
     * Unit test against Issue #315 "Crash when connection could not be established".
     */
    @Test
    fun testMethodErrorResponse() {
        val httpClient = HttpClient(client, "https://demo.test", null, null)

        val host = "just.a.local.url.local"
        try {
            runBlocking {
                httpClient.get("https://$host").asStatus()
            }
            assertTrue("Request should not succeed", true)
        } catch (e: HttpClient.HttpException) {
            assertEquals(500, e.statusCode)
            assertTrue(e.cause is UnknownHostException)
        }
    }
}
